{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Description\n",
    "\n",
    "Tutorial on how to read from the scBaseCamp TileDB-SOMA database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up\n",
    "\n",
    "[tiledbsoma-py](https://pypi.org/project/tiledbsoma/) must be installed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the path to the database (local or cloud)\n",
    "db_uri = \"/scratch/multiomics/nickyoungblut/tiledb-loader/tiledb-soma_GeneFull_Ex50pAS\"\n",
    "#db_uri = \"/home/nickyoungblut/tmp/tiledb/db_bkup/2025-02-10_tiledb-soma_GeneFull_Ex50pAS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load packages\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tiledbsoma\n",
    "import tiledbsoma.io"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Observations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "soma_joinid: int64 not null\n",
      "obs_id: large_string\n",
      "gene_count: int64\n",
      "umi_count: float\n",
      "barcode: large_string\n",
      "SRX_accession: dictionary<values=string, indices=int32, ordered=0>\n",
      "lib_prep: dictionary<values=string, indices=int32, ordered=0>\n",
      "tech_10x: dictionary<values=string, indices=int32, ordered=0>\n",
      "organism: dictionary<values=string, indices=int32, ordered=0>\n",
      "tissue: dictionary<values=string, indices=int32, ordered=0>\n",
      "disease: dictionary<values=string, indices=int32, ordered=0>\n",
      "purturbation: dictionary<values=string, indices=int32, ordered=0>\n",
      "cell_line: dictionary<values=string, indices=int32, ordered=0>\n",
      "czi_collection_id: dictionary<values=string, indices=int32, ordered=0>\n",
      "czi_collection_name: dictionary<values=string, indices=int32, ordered=0>\n"
     ]
    }
   ],
   "source": [
    "# get obs colnames \n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    print(exp.obs.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>obs_id</th>\n",
       "      <th>gene_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>barcode</th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AAACCCAAGGGAGATA_SRX10681588</td>\n",
       "      <td>1677</td>\n",
       "      <td>4085.0</td>\n",
       "      <td>AAACCCAAGGGAGATA</td>\n",
       "      <td>SRX10681588</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>not specified</td>\n",
       "      <td>lean</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>AAACCCAAGGTTGGAC_SRX10681588</td>\n",
       "      <td>1371</td>\n",
       "      <td>3866.0</td>\n",
       "      <td>AAACCCAAGGTTGGAC</td>\n",
       "      <td>SRX10681588</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>not specified</td>\n",
       "      <td>lean</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>AAACCCACAGTGGTGA_SRX10681588</td>\n",
       "      <td>1891</td>\n",
       "      <td>3936.0</td>\n",
       "      <td>AAACCCACAGTGGTGA</td>\n",
       "      <td>SRX10681588</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>human</td>\n",
       "      <td>blood</td>\n",
       "      <td>not specified</td>\n",
       "      <td>lean</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   soma_joinid                        obs_id  gene_count  umi_count  \\\n",
       "0            0  AAACCCAAGGGAGATA_SRX10681588        1677     4085.0   \n",
       "1            1  AAACCCAAGGTTGGAC_SRX10681588        1371     3866.0   \n",
       "2            2  AAACCCACAGTGGTGA_SRX10681588        1891     3936.0   \n",
       "\n",
       "            barcode SRX_accession      lib_prep     tech_10x organism tissue  \\\n",
       "0  AAACCCAAGGGAGATA   SRX10681588  10x_Genomics  3_prime_gex    human  blood   \n",
       "1  AAACCCAAGGTTGGAC   SRX10681588  10x_Genomics  3_prime_gex    human  blood   \n",
       "2  AAACCCACAGTGGTGA   SRX10681588  10x_Genomics  3_prime_gex    human  blood   \n",
       "\n",
       "         disease purturbation       cell_line czi_collection_id  \\\n",
       "0  not specified         lean  not applicable              None   \n",
       "1  not specified         lean  not applicable              None   \n",
       "2  not specified         lean  not applicable              None   \n",
       "\n",
       "  czi_collection_name  \n",
       "0                None  \n",
       "1                None  \n",
       "2                None  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# view all metadata for the first 3 observations\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read()\n",
    "        .concat()\n",
    "        .slice(0,3) \n",
    "        .to_pandas()\n",
    "    )\n",
    "pd.set_option('display.max_columns', 100)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total cells: 154151716\n"
     ]
    }
   ],
   "source": [
    "# count total observations\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    total_cells = (\n",
    "        exp.obs.read(column_names=[\"obs_id\"])\n",
    "        .concat()\n",
    "        .group_by([])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .to_pandas()[\"count_all\"].values[0]\n",
    "    )\n",
    "print(f\"Total cells: {total_cells}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX15016115</td>\n",
       "      <td>148051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX19719427</td>\n",
       "      <td>107903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX15016116</td>\n",
       "      <td>103365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX22915751</td>\n",
       "      <td>100908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX17521046</td>\n",
       "      <td>98235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20289</th>\n",
       "      <td>SRX15483880</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20290</th>\n",
       "      <td>SRX6640133</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20291</th>\n",
       "      <td>SRX21168907</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20292</th>\n",
       "      <td>SRX19326000</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20293</th>\n",
       "      <td>SRX7059415</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20294 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      SRX_accession  count_all\n",
       "0       SRX15016115     148051\n",
       "1       SRX19719427     107903\n",
       "2       SRX15016116     103365\n",
       "3       SRX22915751     100908\n",
       "4       SRX17521046      98235\n",
       "...             ...        ...\n",
       "20289   SRX15483880         47\n",
       "20290    SRX6640133         47\n",
       "20291   SRX21168907         42\n",
       "20292   SRX19326000         12\n",
       "20293    SRX7059415         10\n",
       "\n",
       "[20294 rows x 2 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# observations per dataset\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read(column_names=[\"SRX_accession\"])\n",
    "        .concat()\n",
    "        .group_by([\"SRX_accession\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>115098143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>25183677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>multiome</td>\n",
       "      <td>4778869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>feature_barcoding</td>\n",
       "      <td>2163264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>vdj</td>\n",
       "      <td>2132920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>not_applicable</td>\n",
       "      <td>997406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>other</td>\n",
       "      <td>973283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>489760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cellplex</td>\n",
       "      <td>291407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>247285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>atac</td>\n",
       "      <td>18685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>fixed_rna</td>\n",
       "      <td>7410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tech_10x  count_all\n",
       "0         3_prime_gex  115098143\n",
       "1         5_prime_gex   25183677\n",
       "2            multiome    4778869\n",
       "3   feature_barcoding    2163264\n",
       "4                 vdj    2132920\n",
       "5      not_applicable     997406\n",
       "6               other     973283\n",
       "7                 NaN     489760\n",
       "8            cellplex     291407\n",
       "9                 NaN     247285\n",
       "10               atac      18685\n",
       "11          fixed_rna       7410"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# obs count per 10X Genomics technology\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read(column_names=[\"tech_10x\"])\n",
    "        .concat()\n",
    "        .group_by([\"tech_10x\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>human</td>\n",
       "      <td>80575883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse</td>\n",
       "      <td>69485617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1830849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>489760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  organism  count_all\n",
       "0    human   80575883\n",
       "1    mouse   69485617\n",
       "2      NaN    1830849\n",
       "3      NaN     489760"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# obs count per organism\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read(column_names=[\"organism\"])\n",
    "        .concat()\n",
    "        .group_by([\"organism\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>51452698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>human</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>16149396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human</td>\n",
       "      <td>multiome</td>\n",
       "      <td>2241595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>human</td>\n",
       "      <td>feature_barcoding</td>\n",
       "      <td>1772353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human</td>\n",
       "      <td>vdj</td>\n",
       "      <td>1570771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>human</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>537019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>human</td>\n",
       "      <td>other</td>\n",
       "      <td>521553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>human</td>\n",
       "      <td>cellplex</td>\n",
       "      <td>80411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>human</td>\n",
       "      <td>atac</td>\n",
       "      <td>15988</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  organism           tech_10x  count_all\n",
       "0    human        3_prime_gex   51452698\n",
       "1    human        5_prime_gex   16149396\n",
       "2    human           multiome    2241595\n",
       "3    human  feature_barcoding    1772353\n",
       "4    human                vdj    1570771\n",
       "5    human     not_applicable     537019\n",
       "6    human              other     521553\n",
       "7    human           cellplex      80411\n",
       "8    human               atac      15988"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create metadata query\n",
    "obs_query = tiledbsoma.AxisQuery(value_filter='organism in [\"human\"]')\n",
    "\n",
    "# obs count per 10X Genomics technology for the query results\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.axis_query(\"RNA\", obs_query=obs_query)\n",
    "        .obs()\n",
    "        .concat()\n",
    "        .group_by([\"organism\", \"tech_10x\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>gene_count</th>\n",
       "      <th>umi_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX22369542</td>\n",
       "      <td>14245</td>\n",
       "      <td>272316.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX22369542</td>\n",
       "      <td>14093</td>\n",
       "      <td>272180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX22369542</td>\n",
       "      <td>14056</td>\n",
       "      <td>282893.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX22369542</td>\n",
       "      <td>14018</td>\n",
       "      <td>253654.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX22369542</td>\n",
       "      <td>13895</td>\n",
       "      <td>265843.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46232</th>\n",
       "      <td>SRX22519782</td>\n",
       "      <td>1000</td>\n",
       "      <td>1372.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46233</th>\n",
       "      <td>SRX22519782</td>\n",
       "      <td>1000</td>\n",
       "      <td>1382.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46234</th>\n",
       "      <td>SRX22519782</td>\n",
       "      <td>1000</td>\n",
       "      <td>1648.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46235</th>\n",
       "      <td>SRX22519782</td>\n",
       "      <td>1000</td>\n",
       "      <td>1612.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46236</th>\n",
       "      <td>SRX22519782</td>\n",
       "      <td>1000</td>\n",
       "      <td>1462.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46237 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      SRX_accession  gene_count  umi_count\n",
       "0       SRX22369542       14245   272316.0\n",
       "1       SRX22369542       14093   272180.0\n",
       "2       SRX22369542       14056   282893.0\n",
       "3       SRX22369542       14018   253654.0\n",
       "4       SRX22369542       13895   265843.0\n",
       "...             ...         ...        ...\n",
       "46232   SRX22519782        1000     1372.0\n",
       "46233   SRX22519782        1000     1382.0\n",
       "46234   SRX22519782        1000     1648.0\n",
       "46235   SRX22519782        1000     1612.0\n",
       "46236   SRX22519782        1000     1462.0\n",
       "\n",
       "[46237 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create metadata query\n",
    "obs_query = tiledbsoma.AxisQuery(\n",
    "    value_filter='organism in [\"human\"] and tech_10x in [\"cellplex\"] and gene_count >= 1000'\n",
    ")\n",
    "\n",
    "# gene and umi counts\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.axis_query(\"RNA\", obs_query=obs_query)\n",
    "        .obs()\n",
    "        .concat()\n",
    "        .select([\"SRX_accession\", \"gene_count\", \"umi_count\"])\n",
    "        .sort_by([(\"gene_count\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>organism</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX24745432</td>\n",
       "      <td>human</td>\n",
       "      <td>multiome</td>\n",
       "      <td>24079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX17878115</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>12815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX21370979</td>\n",
       "      <td>human</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>10135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX25986076</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>9395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX25986075</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>8654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRX25986079</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>8353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRX17878109</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>7067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRX21370976</td>\n",
       "      <td>human</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>6711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRX17878114</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>6385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRX21370977</td>\n",
       "      <td>human</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>6372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRX17878113</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>6357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRX25986074</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>5802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRX17878104</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>4741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRX17878105</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>4143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRX17878112</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>3720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRX20300901</td>\n",
       "      <td>mouse</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>3582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRX20300902</td>\n",
       "      <td>mouse</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>3450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRX20300905</td>\n",
       "      <td>mouse</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>3036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRX20300900</td>\n",
       "      <td>mouse</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>2956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRX17878106</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>2894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRX24534638</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>2870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRX17878111</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>2816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRX21370974</td>\n",
       "      <td>human</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>2254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRX21370975</td>\n",
       "      <td>human</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>2139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SRX20300906</td>\n",
       "      <td>mouse</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>2132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SRX20300903</td>\n",
       "      <td>mouse</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>2049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SRX20300899</td>\n",
       "      <td>mouse</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>2045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SRX17878110</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SRX17878107</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SRX20300904</td>\n",
       "      <td>mouse</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>1937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SRX17878108</td>\n",
       "      <td>human</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>1414</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SRX_accession organism     tech_10x  count_all\n",
       "0    SRX24745432    human     multiome      24079\n",
       "1    SRX17878115    human  3_prime_gex      12815\n",
       "2    SRX21370979    human  5_prime_gex      10135\n",
       "3    SRX25986076    human  3_prime_gex       9395\n",
       "4    SRX25986075    human  3_prime_gex       8654\n",
       "5    SRX25986079    human  3_prime_gex       8353\n",
       "6    SRX17878109    human  3_prime_gex       7067\n",
       "7    SRX21370976    human  5_prime_gex       6711\n",
       "8    SRX17878114    human  3_prime_gex       6385\n",
       "9    SRX21370977    human  5_prime_gex       6372\n",
       "10   SRX17878113    human  3_prime_gex       6357\n",
       "11   SRX25986074    human  3_prime_gex       5802\n",
       "12   SRX17878104    human  3_prime_gex       4741\n",
       "13   SRX17878105    human  3_prime_gex       4143\n",
       "14   SRX17878112    human  3_prime_gex       3720\n",
       "15   SRX20300901    mouse  3_prime_gex       3582\n",
       "16   SRX20300902    mouse  3_prime_gex       3450\n",
       "17   SRX20300905    mouse  3_prime_gex       3036\n",
       "18   SRX20300900    mouse  3_prime_gex       2956\n",
       "19   SRX17878106    human  3_prime_gex       2894\n",
       "20   SRX24534638    human  3_prime_gex       2870\n",
       "21   SRX17878111    human  3_prime_gex       2816\n",
       "22   SRX21370974    human  5_prime_gex       2254\n",
       "23   SRX21370975    human  5_prime_gex       2139\n",
       "24   SRX20300906    mouse  3_prime_gex       2132\n",
       "25   SRX20300903    mouse  3_prime_gex       2049\n",
       "26   SRX20300899    mouse  3_prime_gex       2045\n",
       "27   SRX17878110    human  3_prime_gex       2024\n",
       "28   SRX17878107    human  3_prime_gex       1993\n",
       "29   SRX20300904    mouse  3_prime_gex       1937\n",
       "30   SRX17878108    human  3_prime_gex       1414"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter to target accessions\n",
    "\n",
    "## create query\n",
    "target_acc = [\n",
    "    \"SRX24745432\", \"SRX17878115\", \"SRX21370979\", \"SRX25986076\", \"SRX25986075\", \"SRX25986079\", \n",
    "    \"SRX17878109\", \"SRX21370976\", \"SRX21370978\", \"SRX17878114\", \"SRX21370977\", \"SRX17878113\",\n",
    "    \"SRX25986074\", \"SRX17878104\", \"SRX17878105\", \"SRX17878112\", \"SRX20300901\", \"SRX20300902\",\n",
    "    \"SRX20300905\", \"SRX20300900\", \"SRX17878106\", \"SRX24534638\", \"SRX17878111\", \"SRX21370974\",\n",
    "    \"SRX21370975\", \"SRX20300906\", \"SRX20300903\", \"SRX20300899\", \"SRX17878110\", \"SRX17878107\",\n",
    "    \"SRX20300904\", \"SRX17878108\"\n",
    "]\n",
    "target_acc = \",\".join([f'\"{x}\"' for x in target_acc])\n",
    "obs_query = tiledbsoma.AxisQuery(value_filter=f'SRX_accession in [{target_acc}]')\n",
    "\n",
    "## query the database\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.axis_query(\"RNA\", obs_query=obs_query)\n",
    "        .obs()\n",
    "        .concat()\n",
    "        .group_by([\"SRX_accession\", \"organism\", \"tech_10x\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CZI collection datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>150061500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  czi_collection_name  count_all\n",
       "0                None  150061500"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create metadata query\n",
    "obs_query = tiledbsoma.AxisQuery(value_filter='czi_collection_name in [\"None\", \"NaN\"]')\n",
    "\n",
    "# obs count per 10X Genomics technology for the query results\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.axis_query(\"RNA\", obs_query=obs_query)\n",
    "        .obs()\n",
    "        .concat()\n",
    "        .group_by([\"czi_collection_name\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "      <td>189890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "      <td>153916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>139896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Human CellCards Multi-Study CellRef 1.0 Atlas</td>\n",
       "      <td>115615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A spatially resolved atlas of the human lung c...</td>\n",
       "      <td>112047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Single-cell eQTL mapping identifies cell type ...</td>\n",
       "      <td>104481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td>98848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HypoMap ‚Äì a unified single cell gene express...</td>\n",
       "      <td>97187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Multi-scale spatial mapping of cell population...</td>\n",
       "      <td>81460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Cells of the adult human heart</td>\n",
       "      <td>74823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Humoral immunity at the brain borders in homeo...</td>\n",
       "      <td>65517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Inflammatory blockade prevents injury to the d...</td>\n",
       "      <td>63187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Single Cell Multiome Atlas of the Human Fetal ...</td>\n",
       "      <td>51781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Unified comprehensive single-cell atlas of the...</td>\n",
       "      <td>47947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Epithelial zonation along the mammalian small ...</td>\n",
       "      <td>39375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>iPain atlas</td>\n",
       "      <td>35486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>An Integrated Single Cell Atlas of Human Perio...</td>\n",
       "      <td>29639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Transcriptional responses of the human dorsal ...</td>\n",
       "      <td>27222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Integrated adult and foetal heart single-cell ...</td>\n",
       "      <td>26141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Early human lung immune cell development and i...</td>\n",
       "      <td>23776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Single-cell transcriptomics of human T cells r...</td>\n",
       "      <td>22185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Harmonized single-cell landscape, intercellula...</td>\n",
       "      <td>20815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Multimodal single cell sequencing implicates c...</td>\n",
       "      <td>20484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>An alternative cell cycle coordinates multicil...</td>\n",
       "      <td>19902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Single-cell transcriptome profiling of the hum...</td>\n",
       "      <td>18525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Time-resolved Systems Immunology Reveals a Lat...</td>\n",
       "      <td>18419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Single-cell genomics improves the discovery of...</td>\n",
       "      <td>17056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Multi-omic profiling of the developing human c...</td>\n",
       "      <td>16825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Stress-induced RNA‚Äìchromatin interactions pr...</td>\n",
       "      <td>14836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Human Trophoblast Organoids</td>\n",
       "      <td>14210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A transcriptional cross species map of pancrea...</td>\n",
       "      <td>14189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Human skeletal muscle ageing atlas</td>\n",
       "      <td>12222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Integrated scRNA-Seq Identifies Human Postnata...</td>\n",
       "      <td>11416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Single cell transcriptional and chromatin acce...</td>\n",
       "      <td>11173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Spatiotemporal analysis of human intestinal de...</td>\n",
       "      <td>7071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Single-Cell Analysis of Crohn‚Äôs Disease Lesi...</td>\n",
       "      <td>6292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Immune microniches shape intestinal Treg function</td>\n",
       "      <td>4650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>1503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Single-cell transcriptomics of the human retin...</td>\n",
       "      <td>842</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  czi_collection_name  count_all\n",
       "0   Spatial proteogenomics reveals distinct and ev...     189890\n",
       "1   Developmental cell programs are co-opted in in...     153916\n",
       "2   Cross-tissue immune cell analysis reveals tiss...     139896\n",
       "3       Human CellCards Multi-Study CellRef 1.0 Atlas     115615\n",
       "4   A spatially resolved atlas of the human lung c...     112047\n",
       "5   Single-cell eQTL mapping identifies cell type ...     104481\n",
       "6   COVID-19 mRNA vaccine elicits a potent adaptiv...      98848\n",
       "7   HypoMap ‚Äì a unified single cell gene express...      97187\n",
       "8   Multi-scale spatial mapping of cell population...      81460\n",
       "9                      Cells of the adult human heart      74823\n",
       "10  Humoral immunity at the brain borders in homeo...      65517\n",
       "11  Inflammatory blockade prevents injury to the d...      63187\n",
       "12  Single Cell Multiome Atlas of the Human Fetal ...      51781\n",
       "13  Unified comprehensive single-cell atlas of the...      47947\n",
       "14  Epithelial zonation along the mammalian small ...      39375\n",
       "15                                        iPain atlas      35486\n",
       "16  An Integrated Single Cell Atlas of Human Perio...      29639\n",
       "17  Transcriptional responses of the human dorsal ...      27222\n",
       "18  Integrated adult and foetal heart single-cell ...      26141\n",
       "19  Early human lung immune cell development and i...      23776\n",
       "20  Single-cell transcriptomics of human T cells r...      22185\n",
       "21  Harmonized single-cell landscape, intercellula...      20815\n",
       "22  Multimodal single cell sequencing implicates c...      20484\n",
       "23  An alternative cell cycle coordinates multicil...      19902\n",
       "24  Single-cell transcriptome profiling of the hum...      18525\n",
       "25  Time-resolved Systems Immunology Reveals a Lat...      18419\n",
       "26  Single-cell genomics improves the discovery of...      17056\n",
       "27  Multi-omic profiling of the developing human c...      16825\n",
       "28  Stress-induced RNA‚Äìchromatin interactions pr...      14836\n",
       "29                        Human Trophoblast Organoids      14210\n",
       "30  A transcriptional cross species map of pancrea...      14189\n",
       "31                 Human skeletal muscle ageing atlas      12222\n",
       "32  Integrated scRNA-Seq Identifies Human Postnata...      11416\n",
       "33  Single cell transcriptional and chromatin acce...      11173\n",
       "34  Spatiotemporal analysis of human intestinal de...       7071\n",
       "35  Single-Cell Analysis of Crohn‚Äôs Disease Lesi...       6292\n",
       "36  Immune microniches shape intestinal Treg function       4650\n",
       "37  Construction of a human cell landscape at sing...       1503\n",
       "38  Single-cell transcriptomics of the human retin...        842"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create metadata query\n",
    "obs_query = tiledbsoma.AxisQuery(value_filter='czi_collection_name != \"None\"')\n",
    "\n",
    "# obs count per 10X Genomics technology for the query results\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.axis_query(\"RNA\", obs_query=obs_query)\n",
    "        .obs()\n",
    "        .concat()\n",
    "        .group_by([\"czi_collection_name\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1830849"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total counts\n",
    "df[\"count_all\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A spatially resolved atlas of the human lung c...</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Human CellCards Multi-Study CellRef 1.0 Atlas</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Developmental cell programs are co-opted in in...</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cells of the adult human heart</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Humoral immunity at the brain borders in homeo...</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HypoMap ‚Äì a unified single cell gene express...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Early human lung immune cell development and i...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Harmonized single-cell landscape, intercellula...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Unified comprehensive single-cell atlas of the...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Multi-scale spatial mapping of cell population...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>An Integrated Single Cell Atlas of Human Perio...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Inflammatory blockade prevents injury to the d...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>iPain atlas</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Single-cell eQTL mapping identifies cell type ...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Multi-omic profiling of the developing human c...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Single-cell transcriptomics of human T cells r...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Single Cell Multiome Atlas of the Human Fetal ...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Stress-induced RNA‚Äìchromatin interactions pr...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Integrated adult and foetal heart single-cell ...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Multimodal single cell sequencing implicates c...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Transcriptional responses of the human dorsal ...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Single-cell transcriptome profiling of the hum...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>An alternative cell cycle coordinates multicil...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Single cell transcriptional and chromatin acce...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Spatiotemporal analysis of human intestinal de...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Epithelial zonation along the mammalian small ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Integrated scRNA-Seq Identifies Human Postnata...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Human skeletal muscle ageing atlas</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Human Trophoblast Organoids</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Immune microniches shape intestinal Treg function</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Single-cell genomics improves the discovery of...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A transcriptional cross species map of pancrea...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Time-resolved Systems Immunology Reveals a Lat...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Single-Cell Analysis of Crohn‚Äôs Disease Lesi...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Single-cell transcriptomics of the human retin...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  czi_collection_name  count_all\n",
       "0   Cross-tissue immune cell analysis reveals tiss...         52\n",
       "1   A spatially resolved atlas of the human lung c...         34\n",
       "2   Spatial proteogenomics reveals distinct and ev...         27\n",
       "3       Human CellCards Multi-Study CellRef 1.0 Atlas         22\n",
       "4   Developmental cell programs are co-opted in in...         22\n",
       "5                      Cells of the adult human heart         19\n",
       "6   Construction of a human cell landscape at sing...         13\n",
       "7   Humoral immunity at the brain borders in homeo...         12\n",
       "8   COVID-19 mRNA vaccine elicits a potent adaptiv...         11\n",
       "9   HypoMap ‚Äì a unified single cell gene express...          9\n",
       "10  Early human lung immune cell development and i...          8\n",
       "11  Harmonized single-cell landscape, intercellula...          7\n",
       "12  Unified comprehensive single-cell atlas of the...          7\n",
       "13  Multi-scale spatial mapping of cell population...          6\n",
       "14  An Integrated Single Cell Atlas of Human Perio...          5\n",
       "15  Inflammatory blockade prevents injury to the d...          5\n",
       "16                                        iPain atlas          5\n",
       "17  Single-cell eQTL mapping identifies cell type ...          5\n",
       "18  Multi-omic profiling of the developing human c...          4\n",
       "19  Single-cell transcriptomics of human T cells r...          4\n",
       "20  Single Cell Multiome Atlas of the Human Fetal ...          4\n",
       "21  Stress-induced RNA‚Äìchromatin interactions pr...          4\n",
       "22  Integrated adult and foetal heart single-cell ...          3\n",
       "23  Multimodal single cell sequencing implicates c...          3\n",
       "24  Transcriptional responses of the human dorsal ...          3\n",
       "25  Single-cell transcriptome profiling of the hum...          3\n",
       "26  An alternative cell cycle coordinates multicil...          2\n",
       "27  Single cell transcriptional and chromatin acce...          2\n",
       "28  Spatiotemporal analysis of human intestinal de...          2\n",
       "29  Epithelial zonation along the mammalian small ...          2\n",
       "30  Integrated scRNA-Seq Identifies Human Postnata...          2\n",
       "31                 Human skeletal muscle ageing atlas          2\n",
       "32                        Human Trophoblast Organoids          2\n",
       "33  Immune microniches shape intestinal Treg function          2\n",
       "34  Single-cell genomics improves the discovery of...          2\n",
       "35  A transcriptional cross species map of pancrea...          1\n",
       "36  Time-resolved Systems Immunology Reveals a Lat...          1\n",
       "37  Single-Cell Analysis of Crohn‚Äôs Disease Lesi...          1\n",
       "38  Single-cell transcriptomics of the human retin...          1"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of SRX accessions per czi collection\n",
    "\n",
    "# create metadata query\n",
    "obs_query = tiledbsoma.AxisQuery(value_filter='czi_collection_name != \"None\"')\n",
    "\n",
    "# obs count per 10X Genomics technology for the query results\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.axis_query(\"RNA\", obs_query=obs_query)\n",
    "        .obs()\n",
    "        .concat()\n",
    "        .group_by([\"SRX_accession\", \"czi_collection_name\"])\n",
    "        .aggregate([])\n",
    "        .group_by([\"czi_collection_name\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "319"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SRX accession count\n",
    "df[\"count_all\"].sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "soma_joinid: int64 not null\n",
      "var_id: large_string\n"
     ]
    }
   ],
   "source": [
    "# get colnames\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    print(exp.ms[\"RNA\"].var.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>var_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95411</th>\n",
       "      <td>95411</td>\n",
       "      <td>ENSMMUG00000065379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95412</th>\n",
       "      <td>95412</td>\n",
       "      <td>ENSMMUG00000065380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95413</th>\n",
       "      <td>95413</td>\n",
       "      <td>ENSMMUG00000065382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95414</th>\n",
       "      <td>95414</td>\n",
       "      <td>ENSMMUG00000065383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95415</th>\n",
       "      <td>95415</td>\n",
       "      <td>ENSMMUG00000065387</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95416 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid              var_id\n",
       "0                0     ENSG00000000003\n",
       "1                1     ENSG00000000005\n",
       "2                2     ENSG00000000419\n",
       "3                3     ENSG00000000457\n",
       "4                4     ENSG00000000460\n",
       "...            ...                 ...\n",
       "95411        95411  ENSMMUG00000065379\n",
       "95412        95412  ENSMMUG00000065380\n",
       "95413        95413  ENSMMUG00000065382\n",
       "95414        95414  ENSMMUG00000065383\n",
       "95415        95415  ENSMMUG00000065387\n",
       "\n",
       "[95416 rows x 2 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the var level data\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.ms[\"RNA\"]\n",
    "        .var.read(column_names=[\"soma_joinid\", \"var_id\"])\n",
    "        .concat()\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "var_id\n",
       "ENSG       36601\n",
       "ENSMUSG    32285\n",
       "ENSMMUG    26530\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count the number of ensembl IDs per organism\n",
    "df[\"var_id\"].replace(r\"\\d+$\", \"\", regex=True).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Slice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<pyarrow.SparseCOOTensor>\n",
      "type: float\n",
      "shape: (52967109, 68886)\n"
     ]
    }
   ],
   "source": [
    "# sparse count matrix for the first 3 observations\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    print(\n",
    "        exp.ms[\"RNA\"].X[\"data\"]\n",
    "        .read((slice(0,3),))\n",
    "        .coos()\n",
    "        .concat() \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1677, 1371, 1891, 1967, 2050, 1303, 3151, 1705, 1276, 1507],\n",
       "      dtype=int32)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get gene count per obs\n",
    "def get_genes_per_obs(db_uri, start, end):\n",
    "    with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "        data = (\n",
    "            exp.ms[\"RNA\"].X[\"data\"]\n",
    "            .read((slice(start, end),)).coos().concat()\n",
    "        )\n",
    "    return np.diff(data.to_scipy().tocsr().indptr)[start:end]\n",
    "\n",
    "## first 10 observations\n",
    "get_genes_per_obs(db_uri, 0, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([4085., 3866., 3936., 5528., 6013., 2288., 7996., 4050., 3031.,\n",
       "       4726.], dtype=float32)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get UMI count per obs\n",
    "def get_umi_counts_per_obs(db_uri, start, end):\n",
    "    with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "        data = exp.ms[\"RNA\"].X[\"data\"].read((slice(start, end),)).coos().concat()\n",
    "    sp = data.to_scipy().tocsr()\n",
    "    return sp.sum(axis=1).A1[start:end]\n",
    "\n",
    "## first 10 observations\n",
    "get_umi_counts_per_obs(db_uri, 0, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tiledb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
