{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notes\n",
    "\n",
    "* code to check/summarize a tiledb database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import tiledbsoma\n",
    "import tiledbsoma.io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#db_uri = \"/home/nickyoungblut/dev/nextflow/scRecounter/tmp/tiledb/tiledb_exp1\"\n",
    "#db_uri = \"/home/nickyoungblut/dev/nextflow/scRecounter/tmp/tiledb/tiledb_prod3\"\n",
    "#db_uri = \"/home/nickyoungblut/dev/nextflow/scRecounter/tmp/tiledb/tiledb_TEST\"\n",
    "#db_uri = \"/scratch/multiomics/nickyoungblut/tiledb-loader/tiledb_prod_tmp\"\n",
    "#db_uri = \"/scratch/multiomics/nickyoungblut/tiledb-loader/tiledb_prod3_tmp\"\n",
    "#db_uri = \"/scratch/multiomics/nickyoungblut/tiledb-loader/tiledb-soma_GeneFull_Ex50pAS\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "soma_joinid: int64 not null\n",
      "obs_id: large_string\n",
      "gene_count: int64\n",
      "umi_count: float\n",
      "barcode: large_string\n",
      "SRX_accession: dictionary<values=string, indices=int32, ordered=0>\n",
      "lib_prep: dictionary<values=string, indices=int32, ordered=0>\n",
      "tech_10x: dictionary<values=string, indices=int32, ordered=0>\n",
      "organism: dictionary<values=string, indices=int32, ordered=0>\n",
      "tissue: dictionary<values=string, indices=int32, ordered=0>\n",
      "disease: dictionary<values=string, indices=int32, ordered=0>\n",
      "purturbation: dictionary<values=string, indices=int32, ordered=0>\n",
      "cell_line: dictionary<values=string, indices=int32, ordered=0>\n",
      "czi_collection_id: dictionary<values=string, indices=int32, ordered=0>\n",
      "czi_collection_name: dictionary<values=string, indices=int32, ordered=0>\n"
     ]
    }
   ],
   "source": [
    "# get colnames\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    print(exp.obs.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   count_all\n",
      "0   44770961\n"
     ]
    }
   ],
   "source": [
    "# total obs\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    print(\n",
    "        exp.obs.read(column_names=[\"obs_id\"])\n",
    "        .concat()\n",
    "        .group_by([])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .to_pandas()\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5991, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX17521047</td>\n",
       "      <td>93266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX18774274</td>\n",
       "      <td>87650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX18899905</td>\n",
       "      <td>83021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX17917753</td>\n",
       "      <td>78088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX15483882</td>\n",
       "      <td>77392</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SRX_accession  count_all\n",
       "0   SRX17521047      93266\n",
       "1   SRX18774274      87650\n",
       "2   SRX18899905      83021\n",
       "3   SRX17917753      78088\n",
       "4   SRX15483882      77392"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cells per SRX\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read(column_names=[\"obs_id\", \"SRX_accession\"])\n",
    "        .concat()\n",
    "        .group_by([\"SRX_accession\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   organism           tech_10x  count_all\n",
      "0     mouse        3_prime_gex   18278121\n",
      "1     human        3_prime_gex   15413876\n",
      "2     human        5_prime_gex    5574529\n",
      "3     mouse        5_prime_gex    2300447\n",
      "4     mouse           multiome     557445\n",
      "5     human           multiome     523019\n",
      "6     human  feature_barcoding     460394\n",
      "7     human                vdj     438094\n",
      "8     mouse     not_applicable     275526\n",
      "9     human              other     219446\n",
      "10    mouse              other     218551\n",
      "11    human     not_applicable     206333\n",
      "12    mouse                vdj     133023\n",
      "13    mouse           cellplex      77726\n",
      "14    mouse  feature_barcoding      68287\n",
      "15    human           cellplex      26144\n"
     ]
    }
   ],
   "source": [
    "# read in the data and summarize\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    print(\n",
    "        exp.obs.read()\n",
    "        .concat()\n",
    "        .group_by([\"organism\", \"tech_10x\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-02-06 15:20:22.080] [tiledbsoma] [Process: 67865] [Thread: 67865] [warning] [TileDB-SOMA::ManagedQuery] [unnamed] Invalid column selected: gene_symbols\n",
      "(68886, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>var_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   soma_joinid           var_id\n",
       "0            0  ENSG00000000003\n",
       "1            1  ENSG00000000005\n",
       "2            2  ENSG00000000419\n",
       "3            3  ENSG00000000457\n",
       "4            4  ENSG00000000460"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the var (gene) level data\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.ms[\"RNA\"]\n",
    "        .var.read(column_names=[\"soma_joinid\", \"var_id\", \"gene_symbols\"])\n",
    "        .concat()\n",
    "        .to_pandas()\n",
    "    )\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 5 × 68886\n",
      "    obs: 'soma_joinid', 'gene_count', 'umi_count', 'barcode', 'SRX_accession', 'lib_prep', 'tech_10x', 'organism', 'tissue', 'disease', 'purturbation', 'cell_line', 'czi_collection_id', 'czi_collection_name', 'obs_id'\n",
      "    var: 'soma_joinid', 'var_id'\n"
     ]
    }
   ],
   "source": [
    "# slice to just the first 5 cells\n",
    "obs_query = tiledbsoma.AxisQuery(coords=[slice(0, 4)]) \n",
    "\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    adata = exp.axis_query(\"RNA\", obs_query=obs_query).to_anndata(X_name=\"data\")\n",
    "    print(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>count_all</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>263797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>90425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>not_applicable</td>\n",
       "      <td>18041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>multiome</td>\n",
       "      <td>11442</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         tech_10x  count_all\n",
       "0     3_prime_gex     263797\n",
       "1     5_prime_gex      90425\n",
       "2  not_applicable      18041\n",
       "3        multiome      11442"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# obs per tech_10x\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read()\n",
    "        .concat()\n",
    "        .group_by([\"tech_10x\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "print(df.shape)\n",
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene/UMI count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>gene_count</th>\n",
       "      <th>umi_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX18899936</td>\n",
       "      <td>14987</td>\n",
       "      <td>251264.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX18899936</td>\n",
       "      <td>14965</td>\n",
       "      <td>212983.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX18899936</td>\n",
       "      <td>14753</td>\n",
       "      <td>190172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX18899936</td>\n",
       "      <td>14548</td>\n",
       "      <td>181595.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX18899936</td>\n",
       "      <td>14300</td>\n",
       "      <td>171838.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570763</th>\n",
       "      <td>SRX26111881</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570764</th>\n",
       "      <td>SRX26111881</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570765</th>\n",
       "      <td>SRX26111881</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570766</th>\n",
       "      <td>SRX26111881</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570767</th>\n",
       "      <td>SRX26111881</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>570768 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       SRX_accession  gene_count  umi_count\n",
       "0        SRX18899936       14987   251264.0\n",
       "1        SRX18899936       14965   212983.0\n",
       "2        SRX18899936       14753   190172.0\n",
       "3        SRX18899936       14548   181595.0\n",
       "4        SRX18899936       14300   171838.0\n",
       "...              ...         ...        ...\n",
       "570763   SRX26111881           1        4.0\n",
       "570764   SRX26111881           1        5.0\n",
       "570765   SRX26111881           1        6.0\n",
       "570766   SRX26111881           1        4.0\n",
       "570767   SRX26111881           1        5.0\n",
       "\n",
       "[570768 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene and umi counts\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read(column_names=[\"SRX_accession\", \"gene_count\", \"umi_count\"])\n",
    "        .concat()\n",
    "        .sort_by([(\"gene_count\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chris's processed data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_uri = \"/home/nickyoungblut/dev/nextflow/scRecounter/tmp/tiledb/tiledb_counted\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "soma_joinid: int64 not null\n",
      "obs_id: large_string\n",
      "SRX_accession: large_string\n",
      "lib_prep: large_string\n",
      "tech_10x: large_string\n",
      "organism: large_string\n",
      "tissue: large_string\n",
      "disease: large_string\n",
      "purturbation: large_string\n",
      "cell_line: large_string\n",
      "czi_collection_id: large_string\n",
      "czi_collection_name: large_string\n"
     ]
    }
   ],
   "source": [
    "# get colnames\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    print(exp.obs.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>SRX_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGAGATCGC</td>\n",
       "      <td>SRX13549201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGGTCTTTG</td>\n",
       "      <td>SRX13549201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAAGTTGGCGA</td>\n",
       "      <td>SRX13549201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCACAACCCTAA</td>\n",
       "      <td>SRX13549201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCACAATACCCA</td>\n",
       "      <td>SRX13549201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35122</th>\n",
       "      <td>TTTGTTGGTTCAGCTA</td>\n",
       "      <td>SRX21819799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35123</th>\n",
       "      <td>TTTGTTGTCACTGTCC</td>\n",
       "      <td>SRX21819799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35124</th>\n",
       "      <td>TTTGTTGTCGAGTACT</td>\n",
       "      <td>SRX21819799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35125</th>\n",
       "      <td>TTTGTTGTCTAATTCC</td>\n",
       "      <td>SRX21819799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35126</th>\n",
       "      <td>TTTGTTGTCTGTACAG</td>\n",
       "      <td>SRX21819799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35127 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 obs_id SRX_accession\n",
       "0      AAACCCAAGAGATCGC   SRX13549201\n",
       "1      AAACCCAAGGTCTTTG   SRX13549201\n",
       "2      AAACCCAAGTTGGCGA   SRX13549201\n",
       "3      AAACCCACAACCCTAA   SRX13549201\n",
       "4      AAACCCACAATACCCA   SRX13549201\n",
       "...                 ...           ...\n",
       "35122  TTTGTTGGTTCAGCTA   SRX21819799\n",
       "35123  TTTGTTGTCACTGTCC   SRX21819799\n",
       "35124  TTTGTTGTCGAGTACT   SRX21819799\n",
       "35125  TTTGTTGTCTAATTCC   SRX21819799\n",
       "35126  TTTGTTGTCTGTACAG   SRX21819799\n",
       "\n",
       "[35127 rows x 2 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the data\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read(column_names=[\"obs_id\", \"SRX_accession\"])\n",
    "        .concat()\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SRX_accession\n",
       "SRX13549201    14392\n",
       "SRX21819799     7139\n",
       "SRX17937428     5427\n",
       "SRX11523721     5288\n",
       "SRX7117648      2881\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"SRX_accession\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGAGATCGC</td>\n",
       "      <td>SRX13549201</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</td>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGGTCTTTG</td>\n",
       "      <td>SRX13549201</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</td>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAAGTTGGCGA</td>\n",
       "      <td>SRX13549201</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</td>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCACAACCCTAA</td>\n",
       "      <td>SRX13549201</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</td>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCACAATACCCA</td>\n",
       "      <td>SRX13549201</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</td>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35122</th>\n",
       "      <td>TTTGTTGGTTCAGCTA</td>\n",
       "      <td>SRX21819799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc</td>\n",
       "      <td>Unified comprehensive single-cell atlas of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35123</th>\n",
       "      <td>TTTGTTGTCACTGTCC</td>\n",
       "      <td>SRX21819799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc</td>\n",
       "      <td>Unified comprehensive single-cell atlas of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35124</th>\n",
       "      <td>TTTGTTGTCGAGTACT</td>\n",
       "      <td>SRX21819799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc</td>\n",
       "      <td>Unified comprehensive single-cell atlas of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35125</th>\n",
       "      <td>TTTGTTGTCTAATTCC</td>\n",
       "      <td>SRX21819799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc</td>\n",
       "      <td>Unified comprehensive single-cell atlas of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35126</th>\n",
       "      <td>TTTGTTGTCTGTACAG</td>\n",
       "      <td>SRX21819799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc</td>\n",
       "      <td>Unified comprehensive single-cell atlas of the...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35127 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 obs_id SRX_accession     tech_10x organism  \\\n",
       "0      AAACCCAAGAGATCGC   SRX13549201  3_prime_gex      NaN   \n",
       "1      AAACCCAAGGTCTTTG   SRX13549201  3_prime_gex      NaN   \n",
       "2      AAACCCAAGTTGGCGA   SRX13549201  3_prime_gex      NaN   \n",
       "3      AAACCCACAACCCTAA   SRX13549201  3_prime_gex      NaN   \n",
       "4      AAACCCACAATACCCA   SRX13549201  3_prime_gex      NaN   \n",
       "...                 ...           ...          ...      ...   \n",
       "35122  TTTGTTGGTTCAGCTA   SRX21819799          NaN      NaN   \n",
       "35123  TTTGTTGTCACTGTCC   SRX21819799          NaN      NaN   \n",
       "35124  TTTGTTGTCGAGTACT   SRX21819799          NaN      NaN   \n",
       "35125  TTTGTTGTCTAATTCC   SRX21819799          NaN      NaN   \n",
       "35126  TTTGTTGTCTGTACAG   SRX21819799          NaN      NaN   \n",
       "\n",
       "                          czi_collection_id  \\\n",
       "0      74e10dc4-cbb2-4605-a189-8a1cd8e44d8c   \n",
       "1      74e10dc4-cbb2-4605-a189-8a1cd8e44d8c   \n",
       "2      74e10dc4-cbb2-4605-a189-8a1cd8e44d8c   \n",
       "3      74e10dc4-cbb2-4605-a189-8a1cd8e44d8c   \n",
       "4      74e10dc4-cbb2-4605-a189-8a1cd8e44d8c   \n",
       "...                                     ...   \n",
       "35122  a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc   \n",
       "35123  a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc   \n",
       "35124  a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc   \n",
       "35125  a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc   \n",
       "35126  a0c84e3f-a5ca-4481-b3a5-ccfda0a81ecc   \n",
       "\n",
       "                                     czi_collection_name  \n",
       "0      Spatial proteogenomics reveals distinct and ev...  \n",
       "1      Spatial proteogenomics reveals distinct and ev...  \n",
       "2      Spatial proteogenomics reveals distinct and ev...  \n",
       "3      Spatial proteogenomics reveals distinct and ev...  \n",
       "4      Spatial proteogenomics reveals distinct and ev...  \n",
       "...                                                  ...  \n",
       "35122  Unified comprehensive single-cell atlas of the...  \n",
       "35123  Unified comprehensive single-cell atlas of the...  \n",
       "35124  Unified comprehensive single-cell atlas of the...  \n",
       "35125  Unified comprehensive single-cell atlas of the...  \n",
       "35126  Unified comprehensive single-cell atlas of the...  \n",
       "\n",
       "[35127 rows x 6 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the data\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.obs.read(column_names=[\"obs_id\", \"SRX_accession\", \"tech_10x\", \"organism\", \"czi_collection_id\", \"czi_collection_name\"])\n",
    "        .concat()\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      tech_10x organism  count\n",
       "0  3_prime_gex      NaN  22700\n",
       "1          NaN      NaN   7139"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count by tech_10x and organism\n",
    "df.groupby([\"tech_10x\", \"organism\"]).size().reset_index(name=\"count\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Edit existing metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_uri = \"/home/nickyoungblut/dev/nextflow/scRecounter/tmp/tiledb/tiledb_TEST\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyarrow as pa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obs metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "soma_joinid: int64 not null\n",
      "obs_id: large_string\n",
      "SRX_accession: large_string\n",
      "lib_prep: large_string\n",
      "tech_10x: large_string\n",
      "organism: large_string\n",
      "tissue: large_string\n",
      "disease: large_string\n",
      "purturbation: large_string\n",
      "cell_line: large_string\n",
      "czi_collection_id: large_string\n",
      "czi_collection_name: large_string\n"
     ]
    }
   ],
   "source": [
    "# get colnames\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    print(exp.obs.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>obs_id</th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AAACCCAAGAGAGGTA</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>AAACCCAAGAGCAAGA</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>AAACCCAAGCTCTTCC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>AAACCCAAGCTGCCAC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>AAACCCAAGTCACGCC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69058</th>\n",
       "      <td>133935</td>\n",
       "      <td>TTTGGTTAGCCTATTG</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69059</th>\n",
       "      <td>133936</td>\n",
       "      <td>TTTGTTGAGGTTTGAA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69060</th>\n",
       "      <td>133937</td>\n",
       "      <td>TTTGTTGCAGATCCTA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69061</th>\n",
       "      <td>133938</td>\n",
       "      <td>TTTGTTGGTAGTGATA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69062</th>\n",
       "      <td>133939</td>\n",
       "      <td>TTTGTTGGTGGATACG</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69063 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid            obs_id SRX_accession      lib_prep     tech_10x  \\\n",
       "0                0  AAACCCAAGAGAGGTA   SRX23386174  10x_Genomics  3_prime_gex   \n",
       "1                1  AAACCCAAGAGCAAGA   SRX23386174  10x_Genomics  3_prime_gex   \n",
       "2                2  AAACCCAAGCTCTTCC   SRX23386174  10x_Genomics  3_prime_gex   \n",
       "3                3  AAACCCAAGCTGCCAC   SRX23386174  10x_Genomics  3_prime_gex   \n",
       "4                4  AAACCCAAGTCACGCC   SRX23386174  10x_Genomics  3_prime_gex   \n",
       "...            ...               ...           ...           ...          ...   \n",
       "69058       133935  TTTGGTTAGCCTATTG   SRX23508713  10x_Genomics  3_prime_gex   \n",
       "69059       133936  TTTGTTGAGGTTTGAA   SRX23508713  10x_Genomics  3_prime_gex   \n",
       "69060       133937  TTTGTTGCAGATCCTA   SRX23508713  10x_Genomics  3_prime_gex   \n",
       "69061       133938  TTTGTTGGTAGTGATA   SRX23508713  10x_Genomics  3_prime_gex   \n",
       "69062       133939  TTTGTTGGTGGATACG   SRX23508713  10x_Genomics  3_prime_gex   \n",
       "\n",
       "           organism              tissue  \\\n",
       "0      Homo sapiens  bronchial brushing   \n",
       "1      Homo sapiens  bronchial brushing   \n",
       "2      Homo sapiens  bronchial brushing   \n",
       "3      Homo sapiens  bronchial brushing   \n",
       "4      Homo sapiens  bronchial brushing   \n",
       "...             ...                 ...   \n",
       "69058  Homo sapiens      DAOY cell line   \n",
       "69059  Homo sapiens      DAOY cell line   \n",
       "69060  Homo sapiens      DAOY cell line   \n",
       "69061  Homo sapiens      DAOY cell line   \n",
       "69062  Homo sapiens      DAOY cell line   \n",
       "\n",
       "                                            disease  \\\n",
       "0      neovascular age-related macular degeneration   \n",
       "1      neovascular age-related macular degeneration   \n",
       "2      neovascular age-related macular degeneration   \n",
       "3      neovascular age-related macular degeneration   \n",
       "4      neovascular age-related macular degeneration   \n",
       "...                                             ...   \n",
       "69058                                     malignant   \n",
       "69059                                     malignant   \n",
       "69060                                     malignant   \n",
       "69061                                     malignant   \n",
       "69062                                     malignant   \n",
       "\n",
       "                                            purturbation  \\\n",
       "0      bronchial cells dissociated from asthmatic chi...   \n",
       "1      bronchial cells dissociated from asthmatic chi...   \n",
       "2      bronchial cells dissociated from asthmatic chi...   \n",
       "3      bronchial cells dissociated from asthmatic chi...   \n",
       "4      bronchial cells dissociated from asthmatic chi...   \n",
       "...                                                  ...   \n",
       "69058                                   in vitro culture   \n",
       "69059                                   in vitro culture   \n",
       "69060                                   in vitro culture   \n",
       "69061                                   in vitro culture   \n",
       "69062                                   in vitro culture   \n",
       "\n",
       "                    cell_line czi_collection_id czi_collection_name  \n",
       "0      bronchial airway cells              None                None  \n",
       "1      bronchial airway cells              None                None  \n",
       "2      bronchial airway cells              None                None  \n",
       "3      bronchial airway cells              None                None  \n",
       "4      bronchial airway cells              None                None  \n",
       "...                       ...               ...                 ...  \n",
       "69058                    DAOY              None                None  \n",
       "69059                    DAOY              None                None  \n",
       "69060                    DAOY              None                None  \n",
       "69061                    DAOY              None                None  \n",
       "69062                    DAOY              None                None  \n",
       "\n",
       "[69063 rows x 12 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# query\n",
    "obs_query = tiledbsoma.AxisQuery(value_filter='organism in [\"Homo sapiens\"]')\n",
    "#obs_query = tiledbsoma.AxisQuery(value_filter='organism in [\"mouse\"]')\n",
    "\n",
    "# get the target records\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.axis_query(\"RNA\", obs_query=obs_query)\n",
    "        .obs()\n",
    "        .concat()\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update metadata\n",
    "df[\"organism\"] = \"Homo sapiens\"\n",
    "#df[\"organism\"] = \"Mus musculus\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the dtypes\n",
    "for colname in [\"czi_collection_id\", \"czi_collection_name\"]:\n",
    "    df[colname] = df[colname].astype(str)\n",
    "df = pa.Table.from_pandas(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update\n",
    "with tiledbsoma.Experiment.open(db_uri, \"w\") as exp:\n",
    "    exp.obs.write(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add a column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>obs_id</th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AAACCCAAGAGAGGTA</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>AAACCCAAGAGCAAGA</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>AAACCCAAGCTCTTCC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>AAACCCAAGCTGCCAC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>AAACCCAAGTCACGCC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133935</th>\n",
       "      <td>133935</td>\n",
       "      <td>TTTGGTTAGCCTATTG</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133936</th>\n",
       "      <td>133936</td>\n",
       "      <td>TTTGTTGAGGTTTGAA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133937</th>\n",
       "      <td>133937</td>\n",
       "      <td>TTTGTTGCAGATCCTA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133938</th>\n",
       "      <td>133938</td>\n",
       "      <td>TTTGTTGGTAGTGATA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133939</th>\n",
       "      <td>133939</td>\n",
       "      <td>TTTGTTGGTGGATACG</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133940 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        soma_joinid            obs_id SRX_accession      lib_prep  \\\n",
       "0                 0  AAACCCAAGAGAGGTA   SRX23386174  10x_Genomics   \n",
       "1                 1  AAACCCAAGAGCAAGA   SRX23386174  10x_Genomics   \n",
       "2                 2  AAACCCAAGCTCTTCC   SRX23386174  10x_Genomics   \n",
       "3                 3  AAACCCAAGCTGCCAC   SRX23386174  10x_Genomics   \n",
       "4                 4  AAACCCAAGTCACGCC   SRX23386174  10x_Genomics   \n",
       "...             ...               ...           ...           ...   \n",
       "133935       133935  TTTGGTTAGCCTATTG   SRX23508713  10x_Genomics   \n",
       "133936       133936  TTTGTTGAGGTTTGAA   SRX23508713  10x_Genomics   \n",
       "133937       133937  TTTGTTGCAGATCCTA   SRX23508713  10x_Genomics   \n",
       "133938       133938  TTTGTTGGTAGTGATA   SRX23508713  10x_Genomics   \n",
       "133939       133939  TTTGTTGGTGGATACG   SRX23508713  10x_Genomics   \n",
       "\n",
       "           tech_10x      organism              tissue  \\\n",
       "0       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "1       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "2       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "3       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "4       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "...             ...           ...                 ...   \n",
       "133935  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "133936  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "133937  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "133938  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "133939  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "\n",
       "                                             disease  \\\n",
       "0       neovascular age-related macular degeneration   \n",
       "1       neovascular age-related macular degeneration   \n",
       "2       neovascular age-related macular degeneration   \n",
       "3       neovascular age-related macular degeneration   \n",
       "4       neovascular age-related macular degeneration   \n",
       "...                                              ...   \n",
       "133935                                     malignant   \n",
       "133936                                     malignant   \n",
       "133937                                     malignant   \n",
       "133938                                     malignant   \n",
       "133939                                     malignant   \n",
       "\n",
       "                                             purturbation  \\\n",
       "0       bronchial cells dissociated from asthmatic chi...   \n",
       "1       bronchial cells dissociated from asthmatic chi...   \n",
       "2       bronchial cells dissociated from asthmatic chi...   \n",
       "3       bronchial cells dissociated from asthmatic chi...   \n",
       "4       bronchial cells dissociated from asthmatic chi...   \n",
       "...                                                   ...   \n",
       "133935                                   in vitro culture   \n",
       "133936                                   in vitro culture   \n",
       "133937                                   in vitro culture   \n",
       "133938                                   in vitro culture   \n",
       "133939                                   in vitro culture   \n",
       "\n",
       "                     cell_line czi_collection_id czi_collection_name  \n",
       "0       bronchial airway cells              None                None  \n",
       "1       bronchial airway cells              None                None  \n",
       "2       bronchial airway cells              None                None  \n",
       "3       bronchial airway cells              None                None  \n",
       "4       bronchial airway cells              None                None  \n",
       "...                        ...               ...                 ...  \n",
       "133935                    DAOY              None                None  \n",
       "133936                    DAOY              None                None  \n",
       "133937                    DAOY              None                None  \n",
       "133938                    DAOY              None                None  \n",
       "133939                    DAOY              None                None  \n",
       "\n",
       "[133940 rows x 12 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read existing data (must read all rows)\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    obs = exp.obs.read().concat().to_pandas()\n",
    "obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>obs_id</th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>TEST</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AAACCCAAGAGAGGTA</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>AAACCCAAGAGCAAGA</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>AAACCCAAGCTCTTCC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>AAACCCAAGCTGCCAC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>AAACCCAAGTCACGCC</td>\n",
       "      <td>SRX23386174</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>bronchial brushing</td>\n",
       "      <td>neovascular age-related macular degeneration</td>\n",
       "      <td>bronchial cells dissociated from asthmatic chi...</td>\n",
       "      <td>bronchial airway cells</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133935</th>\n",
       "      <td>133935</td>\n",
       "      <td>TTTGGTTAGCCTATTG</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133936</th>\n",
       "      <td>133936</td>\n",
       "      <td>TTTGTTGAGGTTTGAA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133937</th>\n",
       "      <td>133937</td>\n",
       "      <td>TTTGTTGCAGATCCTA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133938</th>\n",
       "      <td>133938</td>\n",
       "      <td>TTTGTTGGTAGTGATA</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133939</th>\n",
       "      <td>133939</td>\n",
       "      <td>TTTGTTGGTGGATACG</td>\n",
       "      <td>SRX23508713</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DAOY cell line</td>\n",
       "      <td>malignant</td>\n",
       "      <td>in vitro culture</td>\n",
       "      <td>DAOY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>test_value</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133940 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        soma_joinid            obs_id SRX_accession      lib_prep  \\\n",
       "0                 0  AAACCCAAGAGAGGTA   SRX23386174  10x_Genomics   \n",
       "1                 1  AAACCCAAGAGCAAGA   SRX23386174  10x_Genomics   \n",
       "2                 2  AAACCCAAGCTCTTCC   SRX23386174  10x_Genomics   \n",
       "3                 3  AAACCCAAGCTGCCAC   SRX23386174  10x_Genomics   \n",
       "4                 4  AAACCCAAGTCACGCC   SRX23386174  10x_Genomics   \n",
       "...             ...               ...           ...           ...   \n",
       "133935       133935  TTTGGTTAGCCTATTG   SRX23508713  10x_Genomics   \n",
       "133936       133936  TTTGTTGAGGTTTGAA   SRX23508713  10x_Genomics   \n",
       "133937       133937  TTTGTTGCAGATCCTA   SRX23508713  10x_Genomics   \n",
       "133938       133938  TTTGTTGGTAGTGATA   SRX23508713  10x_Genomics   \n",
       "133939       133939  TTTGTTGGTGGATACG   SRX23508713  10x_Genomics   \n",
       "\n",
       "           tech_10x      organism              tissue  \\\n",
       "0       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "1       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "2       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "3       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "4       3_prime_gex  Homo sapiens  bronchial brushing   \n",
       "...             ...           ...                 ...   \n",
       "133935  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "133936  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "133937  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "133938  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "133939  3_prime_gex  Homo sapiens      DAOY cell line   \n",
       "\n",
       "                                             disease  \\\n",
       "0       neovascular age-related macular degeneration   \n",
       "1       neovascular age-related macular degeneration   \n",
       "2       neovascular age-related macular degeneration   \n",
       "3       neovascular age-related macular degeneration   \n",
       "4       neovascular age-related macular degeneration   \n",
       "...                                              ...   \n",
       "133935                                     malignant   \n",
       "133936                                     malignant   \n",
       "133937                                     malignant   \n",
       "133938                                     malignant   \n",
       "133939                                     malignant   \n",
       "\n",
       "                                             purturbation  \\\n",
       "0       bronchial cells dissociated from asthmatic chi...   \n",
       "1       bronchial cells dissociated from asthmatic chi...   \n",
       "2       bronchial cells dissociated from asthmatic chi...   \n",
       "3       bronchial cells dissociated from asthmatic chi...   \n",
       "4       bronchial cells dissociated from asthmatic chi...   \n",
       "...                                                   ...   \n",
       "133935                                   in vitro culture   \n",
       "133936                                   in vitro culture   \n",
       "133937                                   in vitro culture   \n",
       "133938                                   in vitro culture   \n",
       "133939                                   in vitro culture   \n",
       "\n",
       "                     cell_line czi_collection_id czi_collection_name  \\\n",
       "0       bronchial airway cells              None                None   \n",
       "1       bronchial airway cells              None                None   \n",
       "2       bronchial airway cells              None                None   \n",
       "3       bronchial airway cells              None                None   \n",
       "4       bronchial airway cells              None                None   \n",
       "...                        ...               ...                 ...   \n",
       "133935                    DAOY              None                None   \n",
       "133936                    DAOY              None                None   \n",
       "133937                    DAOY              None                None   \n",
       "133938                    DAOY              None                None   \n",
       "133939                    DAOY              None                None   \n",
       "\n",
       "              TEST  \n",
       "0       test_value  \n",
       "1       test_value  \n",
       "2       test_value  \n",
       "3       test_value  \n",
       "4       test_value  \n",
       "...            ...  \n",
       "133935  test_value  \n",
       "133936  test_value  \n",
       "133937  test_value  \n",
       "133938  test_value  \n",
       "133939  test_value  \n",
       "\n",
       "[133940 rows x 13 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a new column with default values\n",
    "obs[\"TEST\"] = \"test_value\"\n",
    "# ...or remove column\n",
    "#obs.drop(\"TEST\", inplace=True, axis=1)\n",
    "obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update obs\n",
    "with tiledbsoma.Experiment.open(db_uri, \"w\") as exp:\n",
    "    tiledbsoma.io.update_obs(exp, obs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Var metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "soma_joinid: int64 not null\n",
      "var_id: large_string\n",
      "gene_symbols: large_string\n",
      "feature_types: large_string\n"
     ]
    }
   ],
   "source": [
    "# get colnames\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    print(exp.ms[\"RNA\"].var.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(68886, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>var_id</th>\n",
       "      <th>gene_symbols</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   soma_joinid           var_id gene_symbols    feature_types      organism\n",
       "0            0  ENSG00000243485  MIR1302-2HG  Gene Expression  Homo sapiens\n",
       "1            1  ENSG00000237613      FAM138A  Gene Expression  Homo sapiens\n",
       "2            2  ENSG00000186092        OR4F5  Gene Expression  Homo sapiens\n",
       "3            3  ENSG00000238009   AL627309.1  Gene Expression  Homo sapiens\n",
       "4            4  ENSG00000239945   AL627309.3  Gene Expression  Homo sapiens"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the var (gene) level data\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.ms[\"RNA\"]\n",
    "        .var.read()\n",
    "        .concat()\n",
    "        .to_pandas()\n",
    "    )\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>var_id</th>\n",
       "      <th>gene_symbols</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68881</th>\n",
       "      <td>68881</td>\n",
       "      <td>ENSMUSG00000095523</td>\n",
       "      <td>AC124606.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68882</th>\n",
       "      <td>68882</td>\n",
       "      <td>ENSMUSG00000095475</td>\n",
       "      <td>AC133095.2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68883</th>\n",
       "      <td>68883</td>\n",
       "      <td>ENSMUSG00000094855</td>\n",
       "      <td>AC133095.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68884</th>\n",
       "      <td>68884</td>\n",
       "      <td>ENSMUSG00000095019</td>\n",
       "      <td>AC234645.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68885</th>\n",
       "      <td>68885</td>\n",
       "      <td>ENSMUSG00000095041</td>\n",
       "      <td>AC149090.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68886 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid              var_id gene_symbols    feature_types  \\\n",
       "0                0     ENSG00000243485  MIR1302-2HG  Gene Expression   \n",
       "1                1     ENSG00000237613      FAM138A  Gene Expression   \n",
       "2                2     ENSG00000186092        OR4F5  Gene Expression   \n",
       "3                3     ENSG00000238009   AL627309.1  Gene Expression   \n",
       "4                4     ENSG00000239945   AL627309.3  Gene Expression   \n",
       "...            ...                 ...          ...              ...   \n",
       "68881        68881  ENSMUSG00000095523   AC124606.1  Gene Expression   \n",
       "68882        68882  ENSMUSG00000095475   AC133095.2  Gene Expression   \n",
       "68883        68883  ENSMUSG00000094855   AC133095.1  Gene Expression   \n",
       "68884        68884  ENSMUSG00000095019   AC234645.1  Gene Expression   \n",
       "68885        68885  ENSMUSG00000095041   AC149090.1  Gene Expression   \n",
       "\n",
       "           organism  \n",
       "0      Homo sapiens  \n",
       "1      Homo sapiens  \n",
       "2      Homo sapiens  \n",
       "3      Homo sapiens  \n",
       "4      Homo sapiens  \n",
       "...             ...  \n",
       "68881  Mus musculus  \n",
       "68882  Mus musculus  \n",
       "68883  Mus musculus  \n",
       "68884  Mus musculus  \n",
       "68885  Mus musculus  \n",
       "\n",
       "[68886 rows x 5 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if ENSG, add `Homo sapiens`\n",
    "# if ENSMUSG, add `Mus musculus`\n",
    "def ensemble_to_org(gene_id):\n",
    "    if gene_id.startswith('ENSG'):\n",
    "        return 'Homo sapiens'\n",
    "    elif gene_id.startswith('ENSMUSG'):\n",
    "        return 'Mus musculus'\n",
    "    else:\n",
    "        return 'Unknown'\n",
    "\n",
    "df[\"organism\"] = df[\"var_id\"].apply(ensemble_to_org)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update\n",
    "with tiledbsoma.Experiment.open(db_uri, \"w\") as exp:\n",
    "    tiledbsoma.io.update_var(exp, df, measurement_name=\"RNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>var_id</th>\n",
       "      <th>gene_symbols</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36596</th>\n",
       "      <td>36596</td>\n",
       "      <td>ENSG00000277836</td>\n",
       "      <td>AC141272.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36597</th>\n",
       "      <td>36597</td>\n",
       "      <td>ENSG00000278633</td>\n",
       "      <td>AC023491.2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36598</th>\n",
       "      <td>36598</td>\n",
       "      <td>ENSG00000276017</td>\n",
       "      <td>AC007325.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36599</th>\n",
       "      <td>36599</td>\n",
       "      <td>ENSG00000278817</td>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36600</th>\n",
       "      <td>36600</td>\n",
       "      <td>ENSG00000277196</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid           var_id gene_symbols    feature_types  \\\n",
       "0                0  ENSG00000243485  MIR1302-2HG  Gene Expression   \n",
       "1                1  ENSG00000237613      FAM138A  Gene Expression   \n",
       "2                2  ENSG00000186092        OR4F5  Gene Expression   \n",
       "3                3  ENSG00000238009   AL627309.1  Gene Expression   \n",
       "4                4  ENSG00000239945   AL627309.3  Gene Expression   \n",
       "...            ...              ...          ...              ...   \n",
       "36596        36596  ENSG00000277836   AC141272.1  Gene Expression   \n",
       "36597        36597  ENSG00000278633   AC023491.2  Gene Expression   \n",
       "36598        36598  ENSG00000276017   AC007325.1  Gene Expression   \n",
       "36599        36599  ENSG00000278817   AC007325.4  Gene Expression   \n",
       "36600        36600  ENSG00000277196   AC007325.2  Gene Expression   \n",
       "\n",
       "           organism  \n",
       "0      Homo sapiens  \n",
       "1      Homo sapiens  \n",
       "2      Homo sapiens  \n",
       "3      Homo sapiens  \n",
       "4      Homo sapiens  \n",
       "...             ...  \n",
       "36596  Homo sapiens  \n",
       "36597  Homo sapiens  \n",
       "36598  Homo sapiens  \n",
       "36599  Homo sapiens  \n",
       "36600  Homo sapiens  \n",
       "\n",
       "[36601 rows x 5 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# query by organism: obs & var\n",
    "query = tiledbsoma.AxisQuery(value_filter='organism in [\"Homo sapiens\"]')\n",
    "\n",
    "# get the target records\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    df = (\n",
    "        exp.axis_query(\"RNA\", obs_query=query, var_query=query)\n",
    "        .var()\n",
    "        .concat()\n",
    "        .to_pandas()\n",
    "    )\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# #-- Sandbox --#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import tiledbsoma\n",
    "import tiledbsoma.io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_uri = \"~/dev/nextflow/scRecounter/tmp/tiledb/srx3/tiledb-soma\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "soma_joinid: int64 not null\n",
      "obs_id: large_string\n",
      "SRX_accession: dictionary<values=string, indices=int32, ordered=0>\n"
     ]
    }
   ],
   "source": [
    "# get colnames\n",
    "with tiledbsoma.open(db_uri) as exp:\n",
    "    print(exp.obs.schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  SRX_accession\n",
      "0   SRX24099779\n",
      "1   SRX21101392\n",
      "2   SRX20585665\n"
     ]
    }
   ],
   "source": [
    "# distinct values\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    print(\n",
    "        exp.obs.read(column_names=[\"obs_id\", \"SRX_accession\"])\n",
    "        .concat()\n",
    "        .group_by([\"SRX_accession\"])\n",
    "        .aggregate([])\n",
    "        .to_pandas()\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  SRX_accession  count_all\n",
      "0   SRX21101392       6311\n",
      "1   SRX24099779       3688\n",
      "2   SRX20585665        975\n"
     ]
    }
   ],
   "source": [
    "# cells per SRX\n",
    "with tiledbsoma.Experiment.open(db_uri) as exp:\n",
    "    print(\n",
    "        exp.obs.read(column_names=[\"obs_id\", \"SRX_accession\"])\n",
    "        .concat()\n",
    "        .group_by([\"SRX_accession\"])\n",
    "        .aggregate([\n",
    "            ([], 'count_all'),\n",
    "        ])\n",
    "        .sort_by([(\"count_all\", \"descending\")])\n",
    "        .to_pandas()\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tiledb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
